File: tests/lua-tests/src/CocosDenshionTest/CocosDenshionTest.lua
require "AudioEngine" 
local EFFECT_FILE = "effect1.wav"

local MUSIC_FILE = nil 
local targetPlatform = cc.Application:getInstance():getTargetPlatform()
if (cc.PLATFORM_OS_IPHONE == targetPlatform) or (cc.PLATFORM_OS_IPAD == targetPlatform) then
    MUSIC_FILE = "background.caf"
else
    MUSIC_FILE = "background.mp3"
end

local LINE_SPACE = 40

local function CocosDenshionTest()
    local ret = cc.Layer:create()
    local m_pItmeMenu = nil
    local m_tBeginPos = cc.p(0, 0)
    local m_nSoundId = 0

    local testItems = {
        "play background music",
        "stop background music",
        "pause background music",
        "resume background music",
        "rewind background music",
        "is background music playing",
        "play effect",
        "play effect repeatly",
        "stop effect",
        "unload effect",
        "add background music volume",
        "sub background music volume",
        "add effects volume",
        "sub effects volume",
        "pause effect",
        "resume effect",
        "pause all effects",
        "resume all effects",
        "stop all effects"
    }

    local function menuCallback(tag, pMenuItem)
        local nIdx = pMenuItem:getLocalZOrder() - 10000
        -- play background music
        if nIdx ==  0 then
            AudioEngine.playMusic(MUSIC_FILE, true)
        elseif nIdx == 1 then
            -- stop background music
            AudioEngine.stopMusic()
        elseif nIdx == 2 then
            -- pause background music
            AudioEngine.pauseMusic()
        elseif nIdx == 3 then
            -- resume background music
            AudioEngine.resumeMusic()
            -- rewind background music
        elseif nIdx == 4 then
            AudioEngine.rewindMusic()
        elseif nIdx == 5 then
            -- is background music playing
            if AudioEngine.isMusicPlaying () then
                cclog("background music is playing")
            else
                cclog("background music is not playing")
            end
        elseif nIdx == 6 then
            -- play effect
            m_nSoundId = AudioEngine.playEffect(EFFECT_FILE)
        elseif nIdx == 7 then
            -- play effect
            m_nSoundId = AudioEngine.playEffect(EFFECT_FILE, true)
        elseif nIdx == 8 then
            -- stop effect
            AudioEngine.stopEffect(m_nSoundId)
        elseif nIdx == 9 then
            -- unload effect
            AudioEngine.unloadEffect(EFFECT_FILE)
        elseif nIdx == 10 then
            -- add bakcground music volume
            AudioEngine.setMusicVolume(AudioEngine.getMusicVolume() + 0.1)
        elseif nIdx == 11 then
            -- sub backgroud music volume
            AudioEngine.setMusicVolume(AudioEngine.getMusicVolume() - 0.1)
        elseif nIdx == 12 then
            -- add effects volume
            AudioEngine.setEffectsVolume(AudioEngine.getEffectsVolume() + 0.1)
        elseif nIdx == 13 then
            -- sub effects volume
            AudioEngine.setEffectsVolume(AudioEngine.getEffectsVolume() - 0.1)
        elseif nIdx == 14 then
            AudioEngine.pauseEffect(m_nSoundId)
        elseif nIdx == 15 then
            AudioEngine.resumeEffect(m_nSoundId)
        elseif nIdx == 16 then
            AudioEngine.pauseAllEffects()
        elseif nIdx == 17 then
            AudioEngine.resumeAllEffects()
        elseif nIdx == 18 then
            AudioEngine.stopAllEffects()
        end
    end
    -- add menu items for tests
    m_pItmeMenu = cc.Menu:create()

    m_nTestCount = table.getn(testItems)
    local i = 1
    for  i = 1, m_nTestCount do
        local  label = cc.Label:createWithTTF(testItems[i], s_arialPath, 24)
        label:setAnchorPoint(cc.p(0.5, 0.5))
        local  pMenuItem = cc.MenuItemLabel:create(label)
        pMenuItem:registerScriptTapHandler(menuCallback)
        m_pItmeMenu:addChild(pMenuItem, i + 10000 -1)
        pMenuItem:setPosition( cc.p( VisibleRect:center().x, (VisibleRect:top().y - i * LINE_SPACE) ))
    end

    m_pItmeMenu:setContentSize(cc.size(VisibleRect:getVisibleRect().width, (m_nTestCount + 1) * LINE_SPACE))
    m_pItmeMenu:setPosition(cc.p(0, 0))
    ret:addChild(m_pItmeMenu)

    -- preload background music and effect
    AudioEngine.preloadMusic( MUSIC_FILE )
    AudioEngine.preloadEffect( EFFECT_FILE )

    -- set default volume
    AudioEngine.setEffectsVolume(0.5)
    AudioEngine.setMusicVolume(0.5)

    local function onNodeEvent(event)
        if event == "enter" then

        elseif event == "exit" then
            AudioEngine.destroyInstance()
        end
    end

    ret:registerScriptHandler(onNodeEvent)

    local prev = {x = 0, y = 0}
    local function onTouchEvent(eventType, x, y)
        if eventType == "began" then
            prev.x = x
            prev.y = y
            m_tBeginPos = cc.p(x, y)
            return true
        elseif  eventType == "moved" then
            local touchLocation = cc.p(x, y)
            local nMoveY = touchLocation.y - m_tBeginPos.y
            local curPosX, curPosY = m_pItmeMenu:getPosition()
            local curPos = cc.p(curPosX, curPosY)
            local nextPos = cc.p(curPos.x, curPos.y + nMoveY)

            if nextPos.y < 0.0 then
                m_pItmeMenu:setPosition(cc.p(0, 0))
            end

            if nextPos.y > ((m_nTestCount + 1)* LINE_SPACE - VisibleRect:getVisibleRect().height) then
                m_pItmeMenu:setPosition(cc.p(0, ((m_nTestCount + 1)* LINE_SPACE - VisibleRect:getVisibleRect().height)))
            end

            m_pItmeMenu:setPosition(nextPos)
            m_tBeginPos.x = touchLocation.x
            m_tBeginPos.y = touchLocation.y

            prev.x = x
            prev.y = y
        end
    end

    local function onTouchBegan(touch, event)
        local location = touch:getLocation()
        prev.x = location.x
        prev.y = location.y
        m_tBeginPos = location
        return true
    end

    local function onTouchMoved(touch, event)
        local location = touch:getLocation()
        local touchLocation = location
        local nMoveY = touchLocation.y - m_tBeginPos.y
        local curPosX, curPosY = m_pItmeMenu:getPosition()
        local curPos = cc.p(curPosX, curPosY)
        local nextPos = cc.p(curPos.x, curPos.y + nMoveY)

        if nextPos.y < 0.0 then
            m_pItmeMenu:setPosition(cc.p(0, 0))
        end

        if nextPos.y > ((m_nTestCount + 1)* LINE_SPACE - VisibleRect:getVisibleRect().height) then
            m_pItmeMenu:setPosition(cc.p(0, ((m_nTestCount + 1)* LINE_SPACE - VisibleRect:getVisibleRect().height)))
        end

        m_pItmeMenu:setPosition(nextPos)
        m_tBeginPos.x = touchLocation.x
        m_tBeginPos.y = touchLocation.y

        prev.x = location.x
        prev.y = location.y
    end

    local listener = cc.EventListenerTouchOneByOne:create()
    listener:setSwallowTouches(true)

    listener:registerScriptHandler(onTouchBegan,cc.Handler.EVENT_TOUCH_BEGAN )
    listener:registerScriptHandler(onTouchMoved,cc.Handler.EVENT_TOUCH_MOVED )
    local eventDispatcher = ret:getEventDispatcher()
    eventDispatcher:addEventListenerWithSceneGraphPriority(listener, ret)

    return ret
end

function CocosDenshionTestMain()
    cclog("CocosDenshionTestMain")
    local scene = cc.Scene:create()
    scene:addChild(CocosDenshionTest())
    scene:addChild(CreateBackMenuItem())
    return scene
end

File: tests/js-tests/src/CocosDenshionTest/CocosDenshionTest.js
/****************************************************************************
 Copyright (c) 2008-2010 Ricardo Quesada
 Copyright (c) 2011-2012 cocos2d-x.org
 Copyright (c) 2013-2014 Chukong Technologies Inc.

 http://www.cocos2d-x.org

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

var audioEngine = cc.audioEngine;

var MUSIC_FILE = cc.sys.os == cc.sys.OS_WINRT ? "background.wav" : "background.mp3";
var EFFECT_FILE = cc.sys.os == cc.sys.OS_WINRT ? "effect1.wav" : "effect2.mp3";

var _DenshionTests = [
    'Music Test'
];
var DenshionTests = [
    {
        title:"Play Music",
        playFunc:function () {
            return new playMusic();
        }
    },
    {
        title:"Stop Music",
        playFunc:function () {
            return new stopMusic();
        }
    },
    {
        title:"Pause Music",
        playFunc:function () {
            return new pauseMusic();
        }
    },
    {
        title:"Resume Music",
        playFunc:function () {
            return new resumeMusic();
        }
    },
    {
        title:"Rewind Music",
        playFunc:function () {
            return new rewindMusic();
        }
    },
    {
        title:"is Music Playing",
        playFunc:function () {
            return new isMusicPlaying();
        }
    },
    {
        title:"Increase Music Volume",
        playFunc:function () {
            return new addMusicVolume();
        }
    },
    {
        title:"Decrease Music Volume",
        playFunc:function () {
            return new subMusicVolume();
        }
    },
    {
        title:"Play Sound Effect",
        playFunc:function () {
            return new playEffect();
        }
    },
    {
        title:"Repeat Sound Effect",
        playFunc:function () {
            return new playEffectRepeatly();
        }
    },
    {
        title:"Stop Sound Effect",
        playFunc:function () {
            return new stopEffect();
        }
    },
    {
        title:"Unload Sound Effect",
        playFunc:function () {
            return new unloadEffect();
        }
    },
    {
        title:"Increase Sound Effect Volume",
        playFunc:function () {
            return new addEffectsVolume();
        }
    },
    {
        title:"Decrease Sound Effect Volume",
        playFunc:function () {
            return new subEffectsVolume();
        }
    },
    {
        title:"Pause Sound Effect",
        playFunc:function () {
            return new pauseEffect();
        }
    },
    {
        title:"Resume Sound Effect",
        playFunc:function () {
            return new resumeEffect();
        }
    },
    {
        title:"Pause All Sound Effects",
        playFunc:function () {
            return new pauseAllEffects();
        }
    },
    {
        title:"Resume All Sound Effects",
        playFunc:function () {
            return new resumeAllEffects();
        }
    },
    {
        title:"Stop All Sound Effects",
        playFunc:function () {
            return new stopAllEffects();
        }
    }
];

var CocosDenshionTest = cc.LayerGradient.extend({
    _itemMenu:null,
    _beginPos:cc.p(0, 0),
    _testCount:0,
    ctor:function () {
        this._super(cc.color(0, 0, 0, 255), cc.color(148, 80, 120, 255));

        this._itemMenu = new cc.Menu();
        var winSize = director.getWinSize();
        for (var i = 0; i < DenshionTests.length; i++) {
            var label = new cc.LabelTTF(DenshionTests[i].title, "Arial", 24);
            var menuItem = new cc.MenuItemLabel(label, this.onMenuCallback, this);
            this._itemMenu.addChild(menuItem, i + 10000);
            menuItem.x = winSize.width / 2;
            menuItem.y = winSize.height - (i + 1) * LINE_SPACE;
        }
        this._testCount = i;
        this._itemMenu.width = winSize.width;
	    this._itemMenu.height = (this._testCount + 1) * LINE_SPACE;
        this._itemMenu.x = 0;
        this._itemMenu.y = 0;
        this.addChild(this._itemMenu);

        if( 'touches' in cc.sys.capabilities ) {
            cc.eventManager.addListener({
                event: cc.EventListener.TOUCH_ALL_AT_ONCE,
                onTouchesMoved: function (touches, event) {
                    event.getCurrentTarget().moveMenu(touches[0].getDelta());
                }
            }, this);
        } else if ('mouse' in cc.sys.capabilities )
             cc.eventManager.addListener({
                event: cc.EventListener.MOUSE,
                 onMouseMove: function(event){
                     if(event.getButton() == cc.EventMouse.BUTTON_LEFT)
                        event.getCurrentTarget().moveMenu(event.getDelta());
                 }
             }, this);

        // set default volume
        audioEngine.setEffectsVolume(0.5);
        audioEngine.setMusicVolume(0.5);
    },
    onExit:function () {
        this._super();
        audioEngine.stopMusic();
        audioEngine.stopAllEffects();
    },

    onMenuCallback:function (sender) {
        var idx = sender.zIndex - 10000;
        // create the test scene and run it
        var scene = DenshionTests[idx].playFunc();
    },

    moveMenu:function (delta) {
        var newY = this._itemMenu.y + delta.y;

        if (newY < 0)
            newY = 0;

        if (newY > ((DenshionTests.length + 1) * LINE_SPACE - winSize.height))
            newY = ((DenshionTests.length + 1) * LINE_SPACE - winSize.height);

        this._itemMenu.y = newY;
    }
});

var CocosDenshionTestScene = TestScene.extend({
    runThisTest:function () {

        audioEngine = cc.audioEngine;
        var layer = new CocosDenshionTest();
        this.addChild(layer);
        director.runScene(this);
    }
});

var soundId = null;

var playMusic = function () {
    cc.log("play background music");
    var musicFile = MUSIC_FILE;
    if (cc.sys.os == cc.sys.OS_ANDROID) {
        musicFile = "res/"+musicFile;
    }
    audioEngine.playMusic(musicFile, false);
};

var stopMusic = function () {
    cc.log("stop background music");
    audioEngine.stopMusic();
};

var pauseMusic = function () {
    cc.log("pause background music");
    audioEngine.pauseMusic();
};

var resumeMusic = function () {
    cc.log("resume background music");
    audioEngine.resumeMusic();
};

var rewindMusic = function () {
    cc.log("rewind background music");
    audioEngine.rewindMusic();
};

// is background music playing
var isMusicPlaying = function () {
    if (audioEngine.isMusicPlaying()) {
        cc.log("background music is playing");
    }
    else {
        cc.log("background music is not playing");
    }
};

var playEffect = function () {
    cc.log("play effect");
    soundId = audioEngine.playEffect(EFFECT_FILE);
};

var playEffectRepeatly = function () {
    cc.log("play effect repeatly");
    soundId = audioEngine.playEffect(EFFECT_FILE, true);
};

var stopEffect = function () {
    cc.log("stop effect");
    audioEngine.stopEffect(soundId);
};

var unloadEffect = function () {
    cc.log("unload effect");
    audioEngine.unloadEffect(EFFECT_FILE);
};

var addMusicVolume = function () {
    cc.log("add bakcground music volume");
    audioEngine.setMusicVolume(audioEngine.getMusicVolume() + 0.1);
};

var subMusicVolume = function () {
    cc.log("sub backgroud music volume");
    audioEngine.setMusicVolume(audioEngine.getMusicVolume() - 0.1);
};

var addEffectsVolume = function () {
    cc.log("add effects volume");
    audioEngine.setEffectsVolume(audioEngine.getEffectsVolume() + 0.1);
};

var subEffectsVolume = function () {
    cc.log("sub effects volume");
    audioEngine.setEffectsVolume(audioEngine.getEffectsVolume() - 0.1);
};

var pauseEffect = function () {
    cc.log("pause effect");
    audioEngine.pauseEffect(soundId);
};

var resumeEffect = function () {
    cc.log("resume effect");
    audioEngine.resumeEffect(soundId);
};

var pauseAllEffects = function () {
    cc.log("pause all effects");
    audioEngine.pauseAllEffects();
};
var resumeAllEffects = function () {
    cc.log("resume all effects");
    audioEngine.resumeAllEffects();
};
var stopAllEffects = function () {
    cc.log("stop all effects");
    audioEngine.stopAllEffects();
};